File: Programming and Projection Methods/Lab5C#/Lab5/Lab5/controller/Controller.cs
using Lab5.domain;
using Lab5.service;
using Lab5.utils;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Collections.Specialized;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Lab5.controller
{
    public class Controller:Observer
    {
        private Service service;
        private ObservableCollection<Object> rModel;
        private IList<KeyValuePair<ObservableCollection<RBooking>, List<String>>> rbModelList;

        public Controller(Service service)
        {
            this.service = service;
            service.AddObserver(this);
            rModel = new ObservableCollection<Object>();
            rbModelList = new List<KeyValuePair<ObservableCollection<RBooking>, List<String>>>();
            PopulateRideList();
        }

        private void PopulateRideList()
        {
            rModel.Clear();
            service.findAllRides().ToList().ForEach(x => rModel.Add(new { Destinatia=x.Destination, Data=x.Date, Ora=x.Hour, NrLocuriDisponibile=x.NrPlacesAvailable }));
            rModel.CollectionChanged += RModel_CollectionChanged;
        }

        private void RModel_CollectionChanged(object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e)
        {
            //NotifyCollectionChangedEventArgs args = new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Reset);
            //OnCollectionChanged(args);
        }

        public ObservableCollection<Object> getRModel()
        {
            return rModel;
        }

        public ObservableCollection<RBooking> getRBModel(string destination,string date,string hour)
        {
            ObservableCollection<RBooking> rbModel = new ObservableCollection<RBooking>();
            service.findAllRBookings(destination, date, hour).ToList().ForEach(x => rbModel.Add(x));
            rbModelList.Add(new KeyValuePair<ObservableCollection<RBooking>,List<string>>( rbModel, new List<string> { destination, date, hour }));
            return rbModel;
        }

        public User FindUser(String username,String password)
        {
            return service.findUser(username, password);
        }

        public String bookPlaces(Ride r,String cname,int nrplaces)
        {
            return service.bookPlaces(r, new Client(0, cname), nrplaces);
        }

        public IEnumerable<Ride> GetAllRides()
        {
            return service.findAllRides();
        }
        public IEnumerable<RBooking> FindAllRideBookings(string dest,string date,string hour)
        {
            return service.findAllRBookings(dest, date, hour);
        }

        private void Upd(KeyValuePair<ObservableCollection<RBooking>, List<string>> pair)
        {
            pair.Key.Clear();
            service.findAllRBookings(pair.Value.ElementAt(0), pair.Value.ElementAt(1), pair.Value.ElementAt(2))
                .ToList().ForEach(x => pair.Key.Add(x));
        }

        public override void Update()
        {
            this.PopulateRideList();
            this.rbModelList.ToList().ForEach(x => this.Upd(x));
        }
    }
}

File: Programming and Projection Methods/Lab2/Tema2/src/main/java/controller/Controller.java
package controller;

import domain.Booking;
import domain.Client;
import domain.Ride;
import domain.User;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.util.Pair;
import repository.ValidationException;
import service.Service;
import utils.Observer;
import utils.AppEvent;
import utils.RBooking;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Controller implements Observer<AppEvent> {
    private Service service;
    private ObservableList<Ride> rModel;
    private List<Pair<ObservableList<RBooking>, List<String>>> rbModelList;

    public Controller(Service service) {
        this.service = service;
        service.addObserver(this);
        rModel = FXCollections.observableArrayList();
        rbModelList = new ArrayList<>();
        populateRideList();
    }

    private void populateRideList() {
        rModel.clear();
        service.findAllRides().forEach(x -> rModel.add(x));
    }

    public ObservableList<Ride> getRModel() {
        return rModel;
    }

    public ObservableList<RBooking> getRBModel(String destination, String date, String hour) {
        ObservableList<RBooking> rbModel = FXCollections.observableArrayList();
        service.findAllRBookings(destination, date, hour).forEach(x -> rbModel.add(x));
        rbModelList.add(new Pair<>(rbModel,Arrays.asList(destination, date, hour)));
        return rbModel;
    }

    public void findUser(String username, String password) {
        try (Socket socket = new Socket("localhost", 55555)) {
            PrintWriter writer = new PrintWriter(socket.getOutputStream());
            BufferedReader br = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            String s = username + "|" + password + "\n";
            writer.println(s);
            writer.flush();
            String received = br.readLine();
            if(!received.equals("ok"))
                throw new ValidationException(received);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public String bookPlaces(Ride r, String cname, int nrplaces) {
        return service.bookPlaces(r, new Client(0, cname), nrplaces);
    }

    private void upd(Pair<ObservableList<RBooking>,List<String>> pair) {
        pair.getKey().clear();
        service.findAllRBookings(pair.getValue().get(0),pair.getValue().get(1),pair.getValue().get(2))
                .forEach(x->pair.getKey().add(x));
    }

    @Override
    public void update(AppEvent appEvent) {
        switch (appEvent.getAction()) {
            case UPDATE: {
                populateRideList();
                rbModelList.forEach(x -> upd(x));
            }
        }
    }
}
